openwisp_controller/config/api/serializers.py

```diff
@@ -0,0 +1,217 @@
+import ast
+import collections
+import json
+
+from django.db import transaction
+from django.utils.translation import ugettext_lazy as _
+from rest_framework import serializers
+from swapper import load_model
+
+from openwisp_users.api.mixins import FilterSerializerByOrgManaged
+from openwisp_utils.api.serializers import ValidatedModelSerializer
+
+Template = load_model('config', 'Template')
+Vpn = load_model('config', 'Vpn')
+Device = load_model('config', 'Device')
+Config = load_model('config', 'Config')
+TemplateTag = load_model('config', 'TemplateTag')
+Organization = load_model('openwisp_users', 'Organization')
+
+
+class BaseMeta:
+    read_only_fields = ['created', 'modified']
+
+
+class BaseSerializer(FilterSerializerByOrgManaged, ValidatedModelSerializer):
+    pass
+
+
+class TemplateSerializer(BaseSerializer):
+    config = serializers.JSONField(initial={})
+    tags = serializers.StringRelatedField(many=True, read_only=True)
+    default_values = serializers.JSONField(required=False, initial={})
+
+    class Meta(BaseMeta):
+        model = Template
+        fields = [
+            'id',
+            'name',
+            'tags',
+            'organization',
+            'type',
+            'backend',
+            'default',
+            'required',
+            'default_values',
+            'config',
+            'created',
+            'modified',
+        ]
+
+
+class VpnSerializer(BaseSerializer):
+    config = serializers.JSONField(initial={})
+
+    class Meta(BaseMeta):
+        model = Vpn
+        fields = [
+            'id',
+            'name',
+            'host',
+            'organization',
+            'key',
+            'ca',
+            'cert',
+            'backend',
+            'notes',
+            'dh',
+            'config',
+        ]
+
+
+class BaseJsonField(serializers.Field):
+    def to_internal_value(self, data):
+        if type(data) is str:
+            if len(data) < 2:
```

Comment:
This is added to handle the EOF error, try removing this and just send a POST request with an empty form.

Reply:
> ### `OPENWISP_CONTROLLER_API`
> Why have you removed `OPENWISP_CONTROLLER_API`? Can you please add it back?

###  Didn't you asked me to remove:- https://github.com/openwisp/openwisp-controller/pull/386#discussion_r606573472 
### I have added it back. :+1:

---

> ### Context and Config fields
> There's a couple of issues here:
> 
> ![Screenshot from 2021-04-05 20-46-29](https://user-images.githubusercontent.com/841044/113647231-0b88d900-9650-11eb-9a82-49bbc0be68d3.png)

> 1. the context field is not NetJSON format but JSON (the help text is wrong).
> 2. the output shown in that field is not valid JSON because it uses single quotes, can you make sure it's converted to JSON before being shown?

### Please see
- https://github.com/openwisp/openwisp-controller/pull/386#discussion_r606564829
- https://github.com/openwisp/openwisp-controller/pull/386#discussion_r606664394

### We also had a discussion regarding these fields representation(please see other comments and your past reviews on this)
### If we want to represent this field in JSON then we will need to compromise any one of the representations of HTML form or JSON response. (I have explained to you regarding the issues I was facing) Due to all these reasons, I have been representing these fields in this manner.
### These two fields are behaving weird, reason(due to nested data representation, and default JSON field)

Reply:
Hmm... sounds like instead we should validate if data is valid json instead.